wiring/inc/spark_wiring_system.h

```diff
@@ -360,102 +388,194 @@ class SystemClass {
+        return sleep(pins, pinsSize, edgeTriggerMode, edgeTriggerModeSize, flag, s.count());
     }
 
-    static bool on(system_event_t events, void(*handler)(system_event_t, int,void*)) {
-        return !system_subscribe_event(events, reinterpret_cast<system_event_handler_t*>(handler), nullptr);
+    static String deviceID(void) {
+        return spark_deviceID();
     }
 
-    static bool on(system_event_t events, void(*handler)(system_event_t, int)) {
-        return system_subscribe_event(events, reinterpret_cast<system_event_handler_t*>(handler), NULL);
+    static uint16_t buttonPushed(uint8_t button = 0) {
+        return system_button_pushed_duration(button, nullptr);
     }
 
-    static bool on(system_event_t events, void(*handler)(system_event_t)) {
-        return system_subscribe_event(events, reinterpret_cast<system_event_handler_t*>(handler), NULL);
+    template<typename T>
+    static bool on(system_event_t events, void (T::*handler)(system_event_t, int, void*), T* instance) {
+        using namespace std::placeholders;
+        auto callback = std::bind(handler, instance, _1, _2, _3);
+        auto wrapper = [callback](system_event_t event, int data, void* pointer) {
+            callback(event, data, pointer);
+        };
+        return on(events, wrapper);
     }
 
-    static bool on(system_event_t events, void(*handler)()) {
-        return system_subscribe_event(events, reinterpret_cast<system_event_handler_t*>(handler), NULL);
+    template<typename T>
+    static bool on(system_event_t events, void (T::*handler)(system_event_t, int), T* instance) {
+        using namespace std::placeholders;
+        auto callback = std::bind(handler, instance, _1, _2);
+        auto wrapper = [callback](system_event_t event, int data, void* pointer) {
+            callback(event, data);
+        };
+        return on(events, wrapper);
+    }
+
+    template<typename T>
+    static bool on(system_event_t events, void (T::*handler)(system_event_t), T* instance) {
+        using namespace std::placeholders;
+        auto callback = std::bind(handler, instance, _1);
+        auto wrapper = [callback](system_event_t event, int data, void* pointer) {
+            callback(event);
+        };
+        return on(events, wrapper);
+    }
+
+    template<typename T>
+    static bool on(system_event_t events, void (T::*handler)(), T* instance) {
+        auto callback = std::bind(handler, instance);
+        auto wrapper = [callback](system_event_t event, int data, void* pointer) {
+            callback();
+        };
+        return on(events, wrapper);
+    }
+
+    static bool on(system_event_t events, std::function<void(system_event_t, int, void*)> handler) {
+        // Copy the function object to heap
+        if (!handler) {
+            return -1;
+        }
+        auto wrapper = new std::function<void(system_event_t, int, void*)>(handler); // This may leak memory.
```

Comment:
I think the memory allocated for the wrapper object has to be leaked, since the pointer to the wrapper object that is allocated in heap is not passed to the user as an argument of the `System::off(void(*handler)(system_event_t, int, void*))`. It is possible to unsubscribe events for given handler previously, since the previous handler is a static function which we can compare the function pointer to unsubscribe.

Reply:
We discussed this out-of-band, but will still add a comment here: it's entirely possible not to leak as long as we pass a pointer to the wrapper when registering an event handler and a function pointer that can take care of destroying the opaque (from the system point of view) wrapper object. The deleter function then would just need to cast to the appropriate type and perform destruction with `delete`.